File: 993/build_matrix.py
def build_linear_representation(V):
    # Determine the basis from subsequences V[2^e n + a]
    L = 500
    p = 1000000007 # We can use modulo arithmetic to find the basis index
    
    # We will compute things exactly later, but first find basis indices using mod p
    # Store subsequences
    seqs = []
    # seq_info = (e, a)
    seq_info = []
    
    # To handle V[N]+2N type affine stuff, include 1 and n. 
    # But since BB(N) is k-regular we can just use 1 and V itself.
    # Wait, the definition of 2-regular includes affine terms if we include them. 
    # Let's just use pure subsequences V[2^e n + a], it's enough because it spanned rank 119.
    
    # Actually, we should include the sequence [1, 1, 1...] to handle constants if it helps.
    # Wait, V[2^e n + a] itself is a subspace.
    
    # We need to find exactly rank=119 vectors.
    max_e = 7
    rank = 0
    basis_indices = []
    
    matrix = []
    
    # helper for rref to check if vector is independent
    current_basis_rows = []
    
    def is_independent(vec):
        # check if vec is independent of current_basis_rows mod p
        # just copy current_basis_rows, append vec, and do row reduction
        mat = [row[:] for row in current_basis_rows]
        mat.append(vec[:])
        
        rows = len(mat)
        cols = len(mat[0])
        r = 0
        for c in range(cols):
            pivot = r
            while pivot < rows and mat[pivot][c] % p == 0:
                pivot += 1
            if pivot == rows:
                continue
            mat[r], mat[pivot] = mat[pivot], mat[r]
            inv = pow(mat[r][c], p - 2, p)
            for j in range(c, cols):
                mat[r][j] = (mat[r][j] * inv) % p
            for i in range(rows):
                if i != r and mat[i][c] % p != 0:
                    factor = mat[i][c]
                    for j in range(c, cols):
                        mat[i][j] = (mat[i][j] - factor * mat[r][j]) % p
            r += 1
        return r > len(current_basis_rows)

    idx_to_info = {}
    cnt = 0
    
    for e in range(0, max_e + 1):
        for a in range(2**e):
            sub = [V[2**e * n + a] for n in range(L)]
            
            if is_independent(sub):
                current_basis_rows.append(sub)
                basis_indices.append(cnt)
            
            seqs.append(sub)
            seq_info.append((e, a))
            cnt += 1

    print(f"Found rank {len(basis_indices)}")
    
    # Now we need to express the "target" vectors in terms of the basis EXACTLY.
    # The targets are F_i(2n) and F_i(2n+1).
    # F_i(2n) = V[2^{e_i+1} n + a_i] -> which is exactly a vector in our set with e=e_i+1, a=a_i
    # F_i(2n+1) = V[2^{e_i+1} n + a_i + 2^{e_i}] -> an exact vector in our set with e=e_i+1, a=a_i + 2^{e_i}
    
    # We can express them by solving the linear system USING RATIONALS OR INTEGERS.
    # Luckily, all entries in our vectors are integers, and we just need exact integer or rational coefficients.
    return len(basis_indices)

lines = open('bb_values.txt').read().splitlines()
V = [int(l.split()[1]) for l in lines if l.strip()]
build_linear_representation(V)

File: 993/build_matrix.cpp
#include <iostream>
#include <vector>
#include <fstream>
#include <cassert>

using namespace std;

const long long P = 1000000007;

long long power(long long base, long long exp) {
    long long res = 1;
    base %= P;
    while (exp > 0) {
        if (exp % 2 == 1) res = (res * base) % P;
        base = (base * base) % P;
        exp /= 2;
    }
    return res;
}

long long modInverse(long long n) {
    return power(n, P - 2);
}

int main() {
    ifstream fin("bb_values.txt");
    vector<long long> V;
    long long x, y;
    while (fin >> x >> y) {
        if (x != V.size()) {
            cout << "Error reading file!" << endl;
            return 1;
        }
        V.push_back(y);
    }
    fin.close();
    
    int L = 400; // to keep 2^8 * 400 < 100000
    int max_e = 7;
    
    vector<vector<long long>> seqs;
    vector<pair<int, int>> seq_info;
    
    // In Python we added affine terms. To be exactly equivalent and keep rank stable:
    // Actually, V[N] being exact 2-regular implies that affine terms aren't necessary if the rank
    // includes V, but let's safely add the constant 1 sequence manually just in case.
    vector<long long> ones(L, 1);
    seqs.push_back(ones);
    seq_info.push_back({-1, -1}); // dummy marker for 1 sequence
    
    for (int e = 0; e <= max_e; e++) {
        for (int a = 0; a < (1 << e); a++) {
            vector<long long> sub(L);
            for (int n = 0; n < L; n++) {
                sub[n] = (V[(1<<e)*n + a] % P + P) % P;
            }
            seqs.push_back(sub);
            seq_info.push_back({e, a});
        }
    }
    
    int num_seqs = seqs.size();
    vector<int> basis;
    vector<vector<long long>> row_reduced; // to keep triangular form
    vector<vector<long long>> basis_orig;
    
    vector<vector<long long>> relations(num_seqs);
    
    for (int i = 0; i < num_seqs; i++) {
        vector<long long> cur = seqs[i];
        vector<long long> coeffs(basis.size(), 0);
        
        bool is_indep = false;
        
        for (int j = 0; j < (int)basis.size(); j++) {
            int pivot = -1;
            for(int c=0; c<L; c++) if(row_reduced[j][c] != 0) { pivot = c; break; }
            
            if (pivot != -1 && cur[pivot] != 0) {
                long long factor = (cur[pivot] * modInverse(row_reduced[j][pivot])) % P;
                coeffs[j] = factor;
                for (int c = pivot; c < L; c++) {
                    cur[c] = (cur[c] - factor * row_reduced[j][c]) % P;
                    if (cur[c] < 0) cur[c] += P;
                }
            }
        }
        
        for (int c = 0; c < L; c++) {
            if (cur[c] != 0) {
                is_indep = true;
                break;
            }
        }
        
        if (is_indep) {
            basis.push_back(i);
            row_reduced.push_back(cur);
            basis_orig.push_back(seqs[i]);
            vector<long long> self_rel(basis.size(), 0);
            self_rel.back() = 1;
            relations[i] = self_rel;
        } else {
            relations[i] = coeffs;
        }
    }
    
    cout << "Rank: " << basis.size() << endl;
    
    int d = basis.size();
    vector<vector<long long>> M0(d, vector<long long>(d, 0));
    vector<vector<long long>> M1(d, vector<long long>(d, 0));
    
    bool ok = true;
    for (int i = 0; i < d; i++) {
        int idx = basis[i];
        int e = seq_info[idx].first;
        int a = seq_info[idx].second;
        
        if (e == -1) {
            // This is the sequence of 1s. F(2n) = 1, F(2n+1) = 1
            // So M0 and M1 for this row is just 1 for the 1s basis vector
            // Which idx corresponds to 1s? It's basis[0] (which is idx 0).
            M0[i][0] = 1;
            M1[i][0] = 1;
            continue;
        }
        
        int idx0 = -1, idx1 = -1;
        for (int j=0; j<num_seqs; j++) {
            if (seq_info[j].first == e+1 && seq_info[j].second == a) idx0 = j;
            if (seq_info[j].first == e+1 && seq_info[j].second == a + (1<<e)) idx1 = j;
        }
        
        if (idx0 == -1 || idx1 == -1) {
            cout << "ERROR: basis sequence e=" << e << ", a=" << a << " could not find e+1 derivatives." << endl;
            ok = false;
        } else {
            M0[i] = relations[idx0];
            M1[i] = relations[idx1];
        }
    }
    
    if (!ok) return 1;
    
    ofstream fout("eval.py");
    fout << "P = 1000000007\n";
    fout << "M0 = " << "[\n";
    for(int i=0; i<d; i++) {
        fout << "[";
        for(int j=0; j<d; j++) fout << M0[i][j] << ",";
        fout << "],\n";
    }
    fout << "]\n";
    
    fout << "M1 = " << "[\n";
    for(int i=0; i<d; i++) {
        fout << "[";
        for(int j=0; j<d; j++) fout << M1[i][j] << ",";
        fout << "],\n";
    }
    fout << "]\n";
    
    fout << "F0 = [";
    for(int i=0; i<d; i++) {
        int idx = basis[i];
        if (seq_info[idx].first == -1) fout << 1 << ",";
        else fout << V[seq_info[idx].second] << ",";
    }
    fout << "]\n";
    
    fout.close();
    cout << "Generated eval.py" << endl;
    return 0;
}
